File: src/DuelTest.cs
// Copyright (c) 2020 donaldnevermore
// All rights reserved.
// Licensed under the Apache License, Version 2.0. See the LICENSE file in the project root for more information.

using System.Collections.Generic;
using NUnit.Framework;
using DuelEngine.Domain;
using DuelEngine.MonsterCards;

namespace DuelEngine {
    [TestFixture]
    public class DuelTest {
        private Duel duel;
        private Player player1;
        private Player player2;
        private List<Card> deck1;
        private List<Card> deck2;

        [SetUp]
        public void SetUp() {
            deck1 = new List<Card> {
                new DarkSorcerer(),
                new SpellArcher(),
                new MysticDealer(),

                new DarkSorcerer(),
                new DarkSorcerer(),
                new SpellArcher(),
                new SpellArcher()
            };
            deck2 = new List<Card> {
                new DarkSorcerer(),
                new SpellArcher(),
                new MysticDealer(),

                new DarkSorcerer(),
                new DarkSorcerer(),
                new SpellArcher(),
                new SpellArcher()
            };
            duel = new Duel();
            player1 = new Player { Deck = deck1 };
            player2 = new Player { Deck = deck2 };
            player1.Join(duel);
            player2.Join(duel);

            player1.InitialDraw();
            player2.InitialDraw();

            player1.DrawPhase();
            player1.Draw();
        }

        [Test]
        public void TestSummon() {
            player1.MainPhase();
            var monster = (MonsterCard)player1.Hand[0];
            player1.Summon(monster, 0);
            Assert.AreEqual(monster, player1.MonsterZone[0]);
        }

        [Test]
        public void TestTakeTurns() {
            player1.TurnEnd();
            Assert.AreEqual(2, duel.Turn);

            player2.TurnEnd();
            Assert.AreEqual(3, duel.Turn);
        }

        [Test]
        public void TestHigherAttack() {
            var monster1 = (MonsterCard)player1.Hand[1];
            player1.Summon(monster1, 0);
            player1.TurnEnd();

            var monster2 = (MonsterCard)player2.Hand[0];
            player2.Summon(monster2, 0);
            player2.Attack(player1, 0, 0);

            Assert.AreEqual(7500, player1.LifePoint);
            Assert.AreEqual(null, player1.MonsterZone[0]);
            Assert.AreSame(monster1, player1.Graveyard[0]);
        }

        [Test]
        public void TestLowerAttack() {
            var monster1 = (MonsterCard)player1.Hand[0];
            player1.Summon(monster1, 0);
            player1.TurnEnd();

            var monster2 = (MonsterCard)player2.Hand[1];
            player2.Summon(monster2, 0);
            player2.Attack(player1, 0, 0);

            Assert.AreEqual(7500, player2.LifePoint);
            Assert.AreEqual(null, player2.MonsterZone[0]);
            Assert.AreSame(monster2, player2.Graveyard[0]);
        }

        [Test]
        public void TestMonsterEffect() {
            var mysticDealer = (MonsterCard)player1.Hand[2];
            player1.Summon(mysticDealer, 0);
            // TODO:
            // player1.ActivateEffect(mysticDealer);
            // player1.SelectHandCard(2);
            // player1.HandleEffect(mysticDealer);

            Assert.AreEqual(4, player1.Hand.Count);
        }
    }
}

File: src/lib/duel.test.ts
// Copyright (c) 2020 donaldnevermore
// All rights reserved.
// Licensed under the Apache License, Version 2.0. See the LICENSE file in the project root for more information.

import { Duel } from "./duel"
import { Player } from "./player"
import { Card } from "lib/domain/card"
import { SkyStrikerAceRaye } from "./monster-cards/sky-striker-ace-raye"
import { SkyStrikerAceRoze } from "./monster-cards/sky-striker-ace-roze"
import { Deck } from "./deck"
import { MonsterCard } from "./domain/monster-card"

function setup() {
    const mainDeck1: Card[] = []
    const mainDeck2: Card[] = []

    for (let i = 0; i < 3; i++) {
        mainDeck1.push(new SkyStrikerAceRaye())
        mainDeck2.push(new SkyStrikerAceRaye())
    }
    for (let i = 0; i < 3; i++) {
        mainDeck1.push(new SkyStrikerAceRoze())
        mainDeck2.push(new SkyStrikerAceRoze())
    }

    const p1 = new Player(new Deck(mainDeck1))
    const p2 = new Player(new Deck(mainDeck2))
    const duel = new Duel()
    duel.addPlayers(p1, p2)
    duel.initialize()
    p1.standbyPhase()
    p1.mainPhase1()

    return { p1, p2, duel }
}

test("start", () => {
    const { p1, p2 } = setup()
    expect(p1.hand.length).toBe(5)
    expect(p2.hand.length).toBe(5)
})

test("summon", () => {
    const { p1 } = setup()
    const monster = p1.hand[0] as MonsterCard
    p1.summon(0, 0)
    expect(p1.hand.length).toBe(4)
    expect(monster).toBe(p1.monsterZone[0])
})

test("take turns", () => {
    const { p1, p2, duel } = setup()
    p1.turnEnd()
    expect(duel.turn).toBe(2)
    p2.turnEnd()
    expect(duel.turn).toBe(3)
})

test("should be higher attack", () => {
    const { p1, p2 } = setup()
    const monster1 = p1.hand[3] as MonsterCard
    p1.summon(3, 0)
    p1.turnEnd()

    p2.summon(0, 0)
    p2.attack(p1, 0, 0)

    expect(p1.lifePoints).toBe(7500)
    expect(p1.monsterZone[0]).toBe(undefined)
    expect(p1.graveyard[0]).toBe(monster1)
})

test("should be lower attack", () => {
    const { p1, p2 } = setup()
    p1.summon(0, 0)
    p1.turnEnd()

    const monster2 = p2.hand[3] as MonsterCard
    p2.summon(3, 0)
    p2.attack(p1, 0, 0)

    expect(p2.lifePoints).toBe(7500)
    expect(p2.monsterZone[0]).toBe(undefined)
    expect(p2.graveyard[0]).toBe(monster2)
})

/*
        public void TestMonsterEffect()
        {
            var mysticDealer = (MonsterCard)player1.Hand[2];
            player1.Summon(mysticDealer, 0);
            // TODO:
            // player1.ActivateEffect(mysticDealer);
            // player1.SelectHandCard(2);
            // player1.HandleEffect(mysticDealer);

            Assert.AreEqual(4, player1.Hand.Count);
        }
    }
}
*/
